sw/device/lib/crypto/rsa_3072/rsa_3072_verify.c

```diff
@@ -45,12 +41,85 @@ static const otbn_ptr_t kOtbnVarRsaM0Inv =
+static const uint32_t kOtbnRsaModeNumBytes = sizeof(uint32_t);
 static const uint32_t kOtbnRsaModeConstants = 1;
 static const uint32_t kOtbnRsaModeModexp = 2;
 
-rom_error_t rsa_3072_encode_sha256(const uint8_t *msg, size_t msgLen,
-                                   rsa_3072_int_t *result) {
+/* Exponent is represented by a single word */
+static const uint32_t kOtbnRsaExponentNumBytes = sizeof(uint32_t);
+
+/**
+ * Compute the SHA-256 digest of a message using the HMAC accelerator.
+ *
+ * This function will block if the HMAC fifo fills up, and wait until it can
+ * pass more data.
+ *
+ * `msg` *must* point to an allocated buffer of at least length `msg_len`.
+ *
+ * @param hmac HMAC context
+ * @param msg Message to hash
+ * @param msg_len Length of message in bytes
+ * @param[out] digest Resulting digest (little-endian)
+ */
+dif_result_t sha256(const dif_hmac_t *hmac, const uint8_t *msg, size_t msg_len,
+                    dif_hmac_digest_t *digest) {
+  // HMAC configuration
+  dif_hmac_transaction_t hmac_config = {
+      // Note: the "message_endianness" must always be set to little-endian.
+      // This setting only affects multi-byte writes. Because msg is a byte
+      // array, it will only affect cases where the HMAC driver sees that the
+      // message is word-aligned and decides to push multiple bytes at once
+      // onto the FIFO by converting the first 4 bytes into a 32-bit integer,
+      // which is always little-endian.
+      .message_endianness = kDifHmacEndiannessLittle,
+      .digest_endianness = kDifHmacEndiannessLittle,
+  };
+
+  // Set the HMAC mode to SHA256-only.
+  dif_result_t hmac_result = dif_hmac_mode_sha256_start(hmac, hmac_config);
+  if (hmac_result != kDifOk) {
+    return hmac_result;
+  }
+
+  // Pass the message through to the HMAC device.
+  size_t bytes_sent = 0;
+  while (msg_len > 0) {
+    // Push as many bytes as we can onto the FIFO
+    hmac_result = dif_hmac_fifo_push(hmac, msg, msg_len, &bytes_sent);
+    if ((hmac_result != kDifIpFifoFull) && (hmac_result != kDifOk)) {
+      // Unexpected error; return it
+      return hmac_result;
+    }
+    // Advance message pointer
+    msg = msg + bytes_sent;
+    msg_len = msg_len - bytes_sent;
+  }
+
+  // Start processing
+  hmac_result = dif_hmac_process(hmac);
+  if (hmac_result != kDifOk) {
+    return hmac_result;
+  }
+
+  // Poll until HMAC is finished processing, and retrieve the digest.
+  dif_hmac_digest_t raw_digest;
+  do {
+    hmac_result = dif_hmac_finish(hmac, &raw_digest);
+  } while (hmac_result == kDifUnavailable);
+
+  // Setting the HMAC configuration to little-endian changes the order of bytes
+  // within the words of the digest but not the order of the words themselves;
+  // therefore, to make the digest fully little-endian, we need to reverse the
+  // words.
+  for (uint32_t i = 0; i < ARRAYSIZE(raw_digest.digest); i++) {
+    digest->digest[i] = raw_digest.digest[ARRAYSIZE(raw_digest.digest) - 1 - i];
```

Comment:
Can't you do the reversal in-place and use `digest` above?

I would be happy to have general reversal utilities with names like `memrev8` and `memrev32` in `memory.h`.

Reply:
I thought the code would be simpler this way, but having reversal utilities in `memory.h` is more organized still. Will add those and use them here.

Reply:
Done, mind taking a look at them? They're in their own new commit, so hopefully not too much of a pain to review independently.